pgxn/neon/walproposer.c

```diff
@@ -1281,50 +1234,55 @@ HandleActiveState(Safekeeper *sk, uint32 events)
 {
 	WalProposer *wp = sk->wp;
 
-	uint32		newEvents = WL_SOCKET_READABLE;
-
-	if (events & WL_SOCKET_WRITEABLE)
+	/*
+	 * Note: we don't known which socket awoke us (sk or nwr). However, as
+	 * SendAppendRequests always tries to send at least one msg in
+	 * SS_ACTIVE_SEND be careful not to go there if are only after sk
+	 * response, otherwise it'd create busy loop of pings.
+	 */
+	if (events & WL_SOCKET_WRITEABLE || sk->active_state == SS_ACTIVE_READ_WAL)
 		if (!SendAppendRequests(sk))
 			return;
 
 	if (events & WL_SOCKET_READABLE)
 		if (!RecvAppendResponses(sk))
 			return;
 
-	/*
-	 * We should wait for WL_SOCKET_WRITEABLE event if we have unflushed data
-	 * in the buffer.
-	 *
-	 * LSN comparison checks if we have pending unsent messages. This check
-	 * isn't necessary now, because we always send append messages immediately
-	 * after arrival. But it's good to have it here in case we change this
-	 * behavior in the future.
-	 */
-	if (sk->streamingAt != wp->availableLsn || sk->flushWrite)
-		newEvents |= WL_SOCKET_WRITEABLE;
+#if PG_VERSION_NUM >= 150000
```

Comment:
nit: it's not cool that we have a lot more `#if PG_VERSION_NUM >= XXX` checks now, maybe they can be moved to `walproposer_pg.c`?

Reply:
Well, this is again subtle question of drawing border between walproposer and walproposer_pg. We have three. Two in SafekeeperStateDesiredEvents, called by  AssertEventsOkForState and BlockingWrite; we can move it by creating one more API call. The third is in HandleActiveState, which of course can also avoided by one more API call, but it would look weird.

I'm still not 100% sure what we generally want here, whether logic calculating desired events and setting them should be outside or inside. At least for neon_walreader it easier to handle that mostly in walproposer_pg, it reduces number of api calls.

Reply:
Right, I'm just thinking how to make `walproposer` a library which can be used and tested externally, without dependency on postgres. `PG_VERSION_NUM` is not helping with that :) I don't have a good proposals here yet, just wondering what can we do better